### PR TITLE
support retrying interface exceptions during query execution

### PR DESCRIPTION
**Pull request by colin-rogers-dbt**
requires: https://github.com/dbt-labs/dbt-adapters/pull/368

[docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) dbt-labs/docs.getdbt.com/#

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: dbt/adapters/redshift/connections.py [hunk 1]
```diff
@@ -439,22 +439,18 @@ def open(cls, connection):
 
         credentials = connection.credentials
 
-        def exponential_backoff(attempt: int):
```
Comment (mikealfare) [inline dbt/adapters/redshift/connections.py:442]: We don't need this anymore?

File: hatch.toml
Comment (mikealfare) [inline hatch.toml:1]: Reminder to revert all changes in this file.

File: pyproject.toml
Comment (mikealfare) [inline pyproject.toml:1]: Reminder to revert all changes in this file.

File: tests/unit/test_query.py [hunk 2]
```diff
@@ -99,11 +101,19 @@ def test_execute_without_fetch(self):
 
     def test_add_query_success(self):
         cursor = mock.Mock()
-        with mock.patch.object(SQLConnectionManager, "add_query") as mock_add_query:
+        with (mock.patch.object(SQLConnectionManager, "add_query") as mock_add_query,):
```
Comment (mikealfare) [inline tests/unit/test_query.py:104]: Is this right? It looks like maybe a typo.

Comment (VersusFacit) [reply tests/unit/test_query.py:104]: Artifact of prior code that was there. Cleaned up 🖖

File: dbt/adapters/redshift/connections.py [hunk 3]
```diff
@@ -496,8 +492,18 @@ def add_query(self, sql, auto_begin=True, bindings=None, abridge_sql_log=False):
...
+                redshift_connector.InternalError,
+            )
+
             connection, cursor = super().add_query(
-                query, auto_begin, bindings=bindings, abridge_sql_log=abridge_sql_log
+                query,
+                auto_begin,
+                bindings=bindings,
+                abridge_sql_log=abridge_sql_log,
+                retryable_exceptions=retryable_exceptions,
```
Comment (mikealfare) [inline dbt/adapters/redshift/connections.py:505]: Passing these additional arguments will require a new lower bound on `dbt-adapters`. We don't allow for `**kwargs` in `add_query` (or any of our base methods), so passing a new kwarg will break it.

File: hatch.toml [hunk 4]
```diff
@@ -59,3 +59,6 @@ check-sdist = [
     "pip freeze | grep dbt-redshift",
 ]
 docker-prod = "docker build -f docker/Dockerfile -t dbt-redshift ."
+
+[tool.hatch.metadata]
```
Comment (mikealfare) [inline hatch.toml:63]: This can't be in `hatch.toml` nor `pyproject.toml` to publish to PyPI. `hatch build` will fail since this is not reproducible.

**Review by mikealfare — APPROVED** [review]
Rerunning CI after the version bump is merged into `dbt-adapters` should make it pass.

**Pull request merged**